drivers/input/input_pinnacle.c

```diff
@@ -0,0 +1,827 @@
+		.buf = tx_data,
+		.len = size,
+	}};
+	const struct spi_buf_set tx_set = {
+		.buffers = tx_buf,
+		.count = 1,
+	};
+
+	uint8_t rx_data[size];
+	const struct spi_buf rx_buf[] = {{
+		.buf = rx_data,
+		.len = size,
+	}};
+	const struct spi_buf_set rx_set = {
+		.buffers = rx_buf,
+		.count = 1,
+	};
+
+	int rc;
+
+	rc = spi_transceive_dt(&bus->spi, &tx_set, &rx_set);
+	if (rc) {
+		LOG_ERR("Failed to read from SPI %s", bus->spi.bus->name);
+		return rc;
+	}
+
+	for (i = 0; i < count; ++i) {
+		buf[i] = rx_data[i + 3];
+	}
+
+	return 0;
+}
+#endif /* DT_ANY_INST_ON_BUS_STATUS_OKAY(spi) */
+
+static void pinnacle_decode_sample(const struct device *dev, uint8_t *rx,
+				   union pinnacle_sample *sample)
+{
+	const struct pinnacle_config *config = dev->config;
+
+	if (config->data_mode == PINNACLE_DATA_MODE_ABS) {
+		sample->abs_x = ((rx[2] & 0x0F) << 8) | rx[0];
+		sample->abs_y = ((rx[2] & 0xF0) << 4) | rx[1];
+		sample->abs_z = rx[3] & 0x3F;
+	} else if (config->data_mode == PINNACLE_DATA_MODE_REL) {
+		sample->rel_x = ((rx[0] & BIT(4)) == BIT(4)) ? -(256 - rx[1]) : rx[1];
+		sample->rel_y = ((rx[0] & BIT(5)) == BIT(5)) ? -(256 - rx[2]) : rx[2];
+	}
+}
+
+static bool pinnacle_is_idle_sample(const union pinnacle_sample *sample)
+{
+	return (sample->abs_x == 0 && sample->abs_y == 0 && sample->abs_z == 0);
+}
+
+static void pinnacle_clip_sample(const struct device *dev, union pinnacle_sample *sample)
+{
+	const struct pinnacle_config *config = dev->config;
+
+	if (sample->abs_x < config->active_range_x_min) {
+		sample->abs_x = config->active_range_x_min;
+	}
+	if (sample->abs_x > config->active_range_x_max) {
+		sample->abs_x = config->active_range_x_max;
+	}
+	if (sample->abs_y < config->active_range_y_min) {
+		sample->abs_y = config->active_range_y_min;
+	}
+	if (sample->abs_y > config->active_range_y_max) {
+		sample->abs_y = config->active_range_y_max;
+	}
+}
+
+static void pinnacle_scale_sample(const struct device *dev, union pinnacle_sample *sample)
+{
+	const struct pinnacle_config *config = dev->config;
+
+	sample->abs_x = (uint16_t)((uint32_t)(sample->abs_x - config->active_range_x_min) *
+				   config->resolution_x /
+				   (config->active_range_x_max - config->active_range_x_min));
+	sample->abs_y = (uint16_t)((uint32_t)(sample->abs_y - config->active_range_y_min) *
+				   config->resolution_y /
+				   (config->active_range_y_max - config->active_range_y_min));
+}
+
+static int pinnacle_sample_fetch(const struct device *dev, union pinnacle_sample *sample)
+{
+	const struct pinnacle_config *config = dev->config;
+	const struct pinnacle_bus *bus = &config->bus;
+
+	uint8_t rx[4];
+
+	int ret1 = 0;
+	int ret2 = 0;
+
+	if (config->data_mode == PINNACLE_DATA_MODE_ABS) {
+		ret1 = bus->seq_read(bus, PINNACLE_REG_PACKET_BYTE2, rx, 4);
+	} else if (config->data_mode == PINNACLE_DATA_MODE_REL) {
+		ret1 = bus->seq_read(bus, PINNACLE_REG_PACKET_BYTE0, rx, 3);
+	}
+
+	if (ret1) {
+		LOG_ERR("Failed to read data from SPI device");
```

Comment:
@fabiobaltieri, t he original idea was to clear SW_CC even if we can't read the current state because ASIC waits until an application reads the state and clears SW_CC. Otherwise, it wouldn't do anything until SW_CC. I would appreciate your opinion on whether we at all need to handle this situation in the driver. If not, I would change the code as you suggested.

Reply:
Yeah I understand that but have you seen this happening and working in practice? Normally I'd expect a fail in any `bus->seq_read` to signify a crash of either the device or bus driver and there's not much point even trying to recover from it. Is there any hope that the `bus->seq_read` above fails and the `bus->write` below succeed? I think not.

Reply:
I think I saw something like this at the earlier stage of development of the driver but I've seen this since then. I updated the code as you suggested.